File: arrays and hashing/Top K Frequent Elements.py
import heapq
from collections import defaultdict

def topKFrequent(nums: list[int], k: int) -> list[int]:
    freq = defaultdict(int)
    for num in nums:
        freq[num] += 1

    topK = []

    for num, freq in freq.items():
        heapq.heappush(topK, (-freq, num))

    ans = []
    for _ in range(k):
        ans.append(heapq.heappop(topK)[1])

    return ans

File: arrays and hashing/Top K Frequent Elements.cs
public class SolutionTopKFrequent
{
    public int[] TopKFrequent(int[] nums, int k)
    {
        PriorityQueue<int, int> heap = new PriorityQueue<int, int>();
        Dictionary<int, int> counts = new Dictionary<int, int>();

        foreach (int num in nums)
        {
            if (!counts.ContainsKey(num))
                counts[num] = -1;
            else
                counts[num]--;
        }

        foreach ((int key, int val) in counts)
            heap.Enqueue(key, val);

        int[] ans = new int[k];
        for (int i = 0; i < k; i++)
            ans[i] = heap.Dequeue();

        return ans;
    }
}
